src/audio/dai.c

```diff
@@ -66,6 +93,59 @@ struct dai_data {
 	uint64_t wallclock;	/* wall clock at stream start */
 };
 
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data);
+
+static int dai_add_group(struct comp_dev *dev, uint32_t group_id)
+{
+	int cpu_id = cpu_get_id();
+	struct list_item *dai_groups = &sof_get()->dai_groups[cpu_id].list;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct list_item *group_item;
+	struct dai_group *group;
+
+	if (dd->group) {
+		if (dd->group->group_id != group_id) {
+			comp_err(dev, "dai_add_group(), DAI already in group");
+			return -EINVAL;
+		}
+
+		/* No need to re-assign to the same group, do nothing */
+		return 0;
+	}
+
+	/* Check if this group already exists */
+	list_for_item(group_item, dai_groups) {
+		group = container_of(group_item, struct dai_group, list);
+
+		if (group->group_id == group_id)
+			break;
+
+		group = NULL;
+	}
+
+	/* Group doesn't exist, let's allocate and initialize it */
+	if (!group) {
+		group = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
```

Comment:
This is another reason for dai_group API in lib. dai component code (application layer) should not allocate from the SYS zone.